packages/app/pages/index.tsx

```diff
@@ -201,31 +230,312 @@ const Tab: React.FC<{ to: string }> = ({ children, to }) => {
+          contractInstance.methods
+            .create(name, symbol, quantity, weiPrice, r.data.Hash)
+            .send({ from: address })
+            .on("receipt", (receipt: TransactionReceipt) => {
+              const address =
+                (receipt.events?.["PassportDeployed"]?.returnValues
+                  ?.passport as string) || "";
+              onSuccess({
+                address,
+                symbol,
+                name,
+                supply: Number(quantity),
+                price,
+                metadataHash: r.data.Hash,
+              });
+              resolve();
+            })
+            .on("error", reject)
+        );
+      });
+  }, [
+    symbol,
+    name,
+    quantity,
+    price,
+    contractInstance,
+    web3,
+    address,
+    onSuccess,
+    additionalFields,
+    cid,
+  ]);
+  const stageConfirms = [
+    () => {
+      setStage(1);
+      return true;
+    },
+    () => {
+      setStage(2);
+      return true;
+    },
+    () => {
+      setStage(3);
+      return true;
+    },
+    onFinalConfirm,
+  ];
+  const [fileLoading, setFileLoading] = useState(false);
+  return (
+    <>
+      <Button onClick={open} type="primary" disabled={!address}>
+        Create New Membership Type
+      </Button>
+      <Modal
+        isOpen={isOpen}
+        setIsOpen={setIsOpen}
+        title="New Membership Type"
+        onConfirm={stageConfirms[stage]}
+        confirmText={stage === stageConfirms.length - 1 ? "Create" : "Next"}
+      >
+        {stage === 0 && (
+          <>
+            <ModalInput
+              label={"Name"}
+              value={name}
+              onChange={(e) => setName(e.target.value)}
+            />
+            <ModalInput
+              label={"Symbol"}
+              value={symbol}
+              onChange={(e) => setSymbol(e.target.value)}
+            />
+            <ModalInput
+              label={"Quantity"}
+              value={quantity}
+              onChange={(e) => setQuantity(e.target.value)}
+              type={"number"}
+            />
+            <ModalInput
+              label={"Price"}
+              value={price}
+              onChange={(e) => setPrice(e.target.value)}
+              type={"number"}
+            />
+          </>
+        )}
+        {stage === 1 && (
+          <>
+            <Label label={"Upload Thumbnail"}>
+              <input
+                type={"file"}
+                onChange={async (e) => {
+                  if (e.target.files) {
+                    setFileLoading(true);
+                    const formData = new FormData();
+                    const file = e.target.files[0];
+                    if (file) {
+                      const reader = new FileReader();
+                      reader.onloadend = () => {
+                        const b64 = reader.result as string;
+                        formData.append("files", b64);
+                        return axios
```

Comment:
moving it to `pages/api` would double the runtime, since it would involve uploading a file to the backend, then uploading to IPFS.

If DRY is the goal, I could abstract it out into a utility function